### ISSUE TITLE
Nade Helper

### ISSUE THREAD

**Issue by tewshi0** [CowNowK/AimStar#385]
I was trying to implement nade helper based on AimTux (https://github.com/AimTuxOfficial/AimTux/blob/master/src/Hacks/grenadehelper.cpp)

Here's what I implemented
```
#pragma once
#include "..\MenuConfig.hpp"
#include "..\Entity.h"
#include <vector>
#include <string>
#include <iostream>

namespace NadeHelper
{
	inline DWORD64 GetWeaponHandle(DWORD64 WeaponServices, int WeaponIndex)
	{
		DWORD64 WeaponHandle;
		ProcessMgr.ReadMemory((WeaponServices + Offset::WeaponBaseData.ActiveWeapon) + 0x4 * WeaponIndex, WeaponHandle);

		return WeaponHandle;
	}
	inline DWORD64 GetWeaponServices(const CEntity& LocalPlayer)
	{
		DWORD64 WeaponServices;
		ProcessMgr.ReadMemory(LocalPlayer.Pawn.Address + Offset::Pawn.WeaponServices, WeaponServices);

		return WeaponServices;
	}
	inline int GetWeaponSize(DWORD64 WeaponServices)
	{
		int WeaponSize;
		ProcessMgr.ReadMemory(WeaponServices + Offset::WeaponBaseData.WeaponSize, WeaponSize);

		return WeaponSize;
	}

	float GetDistance3D(Vec3 playerPosition, Vec3 enemyPosition) {
		return sqrt(powf(enemyPosition.x - playerPosition.x, 2.0f)
			+ powf(enemyPosition.y - playerPosition.y, 2.0f)
			+ powf(enemyPosition.z - playerPosition.z, 2.0f));
	}

	void AngleVectors(const Vec3& angles, Vec3* forward)
	{

		float RadY = M_PI / 180.0f * angles.y;
		float RadX = M_PI / 180.0f * angles.x;

		float sp, sy, cp, cy;

		sp = (float)sin(RadX);
		cp = (float)cos(RadX);

		sy = (float)sin(RadY);
		cy = (float)cos(RadY);

		forward->x = (cp * cy);
		forward->y = (cp * sy);
		forward->z = -sp;

	}

    void DrawNadeInfo_bak(CEntity& LocalEntity, CGame Game)
    {
        if (!MiscCFG::NadeHelper)
            return;

		uintptr_t ClippingWeapon, WeaponData, WeaponNameAddress;
		float dist;
		Vec2 pos2d, pos2dAng;
		Vec3 TempLoc, TempAng, PlayerPos;

		PlayerPos = LocalEntity.Pawn.Pos;

		ProcessMgr.ReadMemory(LocalEntity.Pawn.Address + Offset::Pawn.pClippingWeapon, ClippingWeapon);
		ProcessMgr.ReadMemory(ClippingWeapon + Offset::WeaponBaseData.WeaponDataPTR, WeaponData);
		ProcessMgr.ReadMemory(WeaponData + Offset::WeaponBaseData.szName, WeaponNameAddress);

		std::string weaponName = XorStr("Invalid Weapon Name");

		if (!WeaponNameAddress)
		{
			weaponName = XorStr("NULL");
		}
		else {
			weaponName = LocalEntity.Pawn.WeaponName;
		}
		Vec2 HeadPos;

		ImGuiWindowFlags windowFlags = ImGuiWindowFlags_NoCollapse | ImGuiWindowFlags_NoTitleBar | ImGuiWindowFlags_NoResize | ImGuiWindowFlags_AlwaysAutoResize;
		ImGui::SetNextWindowBgAlpha(0.5f);
		ImGui::Begin(XorStr("Nade Helper"), nullptr, windowFlags);
		ImGui::Text(XorStr("Debug Window"));
		ImGui::Text(XorStr("Weapon Name: %s"), LocalEntity.Pawn.WeaponName);
		ImGui::Text(XorStr("AbsPos: %.3f %.3f %.3f"), LocalEntity.Pawn.Pos.x, LocalEntity.Pawn.Pos.y, LocalEntity.Pawn.Pos.z);
		ImGui::Text(XorStr("Viewangle: %.3f %.3f"), LocalEntity.Pawn.ViewAngle.x, LocalEntity.Pawn.ViewAngle.y);

		if (weaponName == "smokegrenade") {
			//Throw from
			TempLoc.x = -1055.31;
			TempLoc.y = 2553.84;
			TempLoc.z = 24.4448;


			dist = GetDistance3D(PlayerPos, TempLoc);
			if (dist < 800) {
				if (gGame.View.WorldToScreen(TempLoc, pos2d)) {
					Gui.CircleFilled(Vec2(pos2d.x, pos2d.y), 6, ImColor(0, 0, 0));
					Gui.CircleFilled(Vec2(pos2d.x, pos2d.y), 4, ImColor(255, 255, 0));
					Gui.StrokeText("Top Mid", { pos2d.x + 8, pos2d.y - 6 }, ImColor(255, 204, 0, 255), 14, false);
				}

				dist = GetDistance3D(PlayerPos, TempLoc);

				if (dist <= 5) {
					Gui.CircleFilled(Vec2(pos2d.x, pos2d.y), 6, ImColor(0, 0, 0));
					Gui.CircleFilled(Vec2(pos2d.x, pos2d.y), 4, ImColor(255, 255, 0));
					//Look at
					Vec3 infoVec = Vec3(-653.984, 1746.71, 512.222);

					Vec2 posVec;

					if (!gGame.View.WorldToScreenNade(infoVec, posVec)) {
						return;
					}

					Vec2 pos2d(posVec.x, posVec.y);

					Gui.CircleFilled(Vec2(pos2d.x, pos2d.y), 4, ImColor(0, 0, 0));
					Gui.CircleFilled(Vec2(pos2d.x, pos2d.y), 3, ImColor(255, 255, 0));

				}
			}
		}
		ImGui::Text(XorStr("Distance: %.3f"), dist);

		ImGui::End();
    }
}
```

However I had to modify WorldToScreen function, for some reason the calculated value for LookAt was <0.01 (negative) which means object is off screen. This lead to nothing being drawn on the screen. I fixed this by creating a copy of WorldToScreen function and removing

```
if (View <= 0.01)
	return false;
```
This has made the nade helper fuctional but it now draws two points one in the intended location and other one directly opposite (behind/under player). I have looked through multiple codebases and all of them use same WorldToScreen with check for negative value. I am unable to understand why it doesn't work here.

**Comment by M3351AN** [comment]
```
if (View <= 0.01)
	return false;
```
this is depth chk of matrics calculate, if u del it, w2s fuction will calculate a screen pos with additive inverse(opposite) value, thats why it would render a other one point that directly opposite 
i will take a look at the nade helper later

**Comment by tewshi0** [comment]
I decided to have another look at it and turns out all I had to do was adjust value for depth check to be more relaxed. Using value 0.001f (or even 0) for depth check fixed the issue for me. I converted the grenade data from orbit to json and updated my function to load it. It works pretty well for now, need to implement some sort of grenade recorder and line up manager so I can easily clean up the nade data.
